src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+}
+
+// duplicate a handle to the target process, closing the source handle
+fn move_handle_to_process(handle: HANDLE, other_process: HANDLE) -> Result<HANDLE,WinError> {
+    dup_handle_to_process_with_flags(handle, other_process, winapi::DUPLICATE_CLOSE_SOURCE | winapi::DUPLICATE_SAME_ACCESS)
+}
+
+// get the current process handle as a handle in another process
+#[allow(dead_code)]
+fn current_process_handle_to_process(other_process: HANDLE) -> Result<HANDLE,WinError> {
+    unsafe {
+        let mut new_handle: HANDLE = INVALID_HANDLE_VALUE;
+        let ok = kernel32::DuplicateHandle(kernel32::GetCurrentProcess(), kernel32::GetCurrentProcess(),
+                                           other_process, &mut new_handle,
+                                           winapi::PROCESS_DUP_HANDLE, winapi::FALSE, 0);
+        if ok == winapi::FALSE {
+            Err(WinError::last("DuplicateHandle"))
+        } else {
+            Ok(new_handle)
+        }
+    }
+}
+
+fn create_overlapped() -> Box<winapi::OVERLAPPED> {
+    Box::new(winapi::OVERLAPPED {
+        Internal: 0,
+        InternalHigh: 0,
+        Offset: 0,
+        OffsetHigh: 0,
+        hEvent: 0 as HANDLE
+    })
+}
+
+fn reset_overlapped(ov: &mut winapi::OVERLAPPED) {
+    ov.Internal = 0;
+    ov.InternalHigh = 0;
+    ov.Offset = 0;
+    ov.OffsetHigh = 0;
+    ov.hEvent = 0 as HANDLE;
+}
+
+
+#[derive(Debug, PartialEq)]
+struct WinHandle {
+    h: HANDLE
+}
+
+unsafe impl Send for WinHandle { }
+unsafe impl Sync for WinHandle { }
+
+impl Drop for WinHandle {
+    fn drop(&mut self) {
+        self.close();
+    }
+}
+
+impl Default for WinHandle {
+    fn default() -> WinHandle {
+        WinHandle { h: INVALID_HANDLE_VALUE }
+    }
+}
+
+impl Deref for WinHandle {
+    type Target = HANDLE;
+
+    #[inline]
+    fn deref(&self) -> &HANDLE {
+        &self.h
+    }
+}
+
+impl WinHandle {
+    fn new(h: HANDLE) -> WinHandle {
+        WinHandle { h: h }
+    }
+
+    #[allow(dead_code)]
+    fn invalid() -> WinHandle {
+        WinHandle { h: INVALID_HANDLE_VALUE }
+    }
+
+    fn is_valid(&self) -> bool {
+        self.h != INVALID_HANDLE_VALUE
+    }
+
+    fn close(&mut self) {
+        unsafe {
+            dd!("WinHandle drop: {:?}", self.h);
+            let h = self.take();
+            kernel32::CloseHandle(h);
+        }
+    }
+
+    fn take(&mut self) -> HANDLE {
+        mem::replace(&mut self.h, INVALID_HANDLE_VALUE)
+    }
+}
+
+// MessageReader implements blocking/nonblocking reads of messages
+// from the handle
+#[derive(Debug)]
+struct MessageReader {
+    // The reader handle; same as the one on OsIpcReceiver, here for
+    // convenience
+    handle: HANDLE,
+
+    // The OVERLAPPED struct for async IO on this receiver; we'll only
+    // ever have one in flight
+    ov: Box<winapi::OVERLAPPED>,
+
+    // A read buffer for any pending reads
+    read_buf: Vec<u8>,
+
+    // If we have already issued an async read
+    read_in_progress: bool,
```

Comment:
This actually started as a state machine with an enum tracking the current state.  I eventually transformed the code down so that there were ever only two states, one that we had a read in flight, and one that we didn't.  So I dropped the enum and made it a bool instead, because that made more sense -- it's simpler to write `if self.read_in_progress` than `if self.state == State::ReadInProgress`, and much simpler than a `match` with two arms.